File: src/main/java/com/fengsheng/handler/choose_whether_receive_tos.java
package com.fengsheng.handler;

import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.phase.MessageMoveNext;
import com.fengsheng.phase.OnChooseReceiveCard;
import com.fengsheng.phase.SendPhaseIdle;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class choose_whether_receive_tos extends AbstractProtoHandler<Fengsheng.choose_whether_receive_tos> {
    private static final Logger log = Logger.getLogger(choose_whether_receive_tos.class);

    @Override
    protected void handle0(HumanPlayer r, Fengsheng.choose_whether_receive_tos pb) {
        if (!r.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        if (!(r.getGame().getFsm() instanceof SendPhaseIdle fsm) || r != fsm.inFrontOfWhom) {
            log.error("不是选择是否接收情报的时机");
            return;
        }
        if (pb.getReceive()) {
            r.incrSeq();
            r.getGame().resolve(new OnChooseReceiveCard(fsm.whoseTurn, fsm.messageCard, fsm.inFrontOfWhom, fsm.isMessageCardFaceUp));
        } else {
            if (r == fsm.whoseTurn) {
                log.error("传出者必须接收");
                return;
            }
            boolean locked = false;
            for (Player a : fsm.lockedPlayers) {
                if (r == a) {
                    locked = true;
                    break;
                }
            }
            if (locked) {
                log.error("被锁定，必须接收");
                return;
            }
            r.incrSeq();
            r.getGame().resolve(new MessageMoveNext(fsm));
        }
    }
}

File: src/main/kotlin/handler/choose_whether_receive_tos.kt
package com.fengsheng.handler

import com.fengsheng.HumanPlayer
import com.fengsheng.phase.MessageMoveNext
import com.fengsheng.phase.OnChooseReceiveCard
import com.fengsheng.phase.SendPhaseIdle
import com.fengsheng.protos.Fengsheng
import com.fengsheng.skill.cannotReceiveMessage
import com.fengsheng.skill.mustReceiveMessage
import org.apache.logging.log4j.kotlin.logger

class choose_whether_receive_tos : AbstractProtoHandler<Fengsheng.choose_whether_receive_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.choose_whether_receive_tos) {
        if (!r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        val fsm = r.game!!.fsm as? SendPhaseIdle
        if (r !== fsm?.inFrontOfWhom) {
            logger.error("不是选择是否接收情报的时机")
            r.sendErrorMessage("不是选择是否接收情报的时机")
            return
        }
        val mustReceive = fsm.mustReceiveMessage()
        if (pb.receive) {
            if (!mustReceive && fsm.cannotReceiveMessage()) {
                logger.error("不能选择接收情报")
                r.sendErrorMessage("不能选择接收情报")
                return
            }
            r.incrSeq()
            r.game!!.resolve(
                OnChooseReceiveCard(
                    fsm.whoseTurn,
                    fsm.sender,
                    fsm.messageCard,
                    fsm.inFrontOfWhom,
                    fsm.isMessageCardFaceUp
                )
            )
        } else {
            if (mustReceive) {
                logger.error("必须选择接收情报")
                r.sendErrorMessage("必须选择接收情报")
                return
            }
            r.incrSeq()
            r.game!!.resolve(MessageMoveNext(fsm))
        }
    }
}